torch/utils/tensorboard/_pytorch_graph.py

```diff
@@ -0,0 +1,267 @@
+            if type(node) == NodeBase:
+                self.uniqueNameToScopedName[key] = node.scope + '/' + node.uniqueName
+            if hasattr(node, 'input_or_output'):
+                self.uniqueNameToScopedName[key] = node.input_or_output + '/' + node.uniqueName
+            if hasattr(node, 'scope'):
+                if node.scope == '' and self.shallowestScopeName:
+                    self.uniqueNameToScopedName[node.uniqueName] = self.shallowestScopeName + '/' + node.uniqueName
+
+        # replace name
+        for key, node in self.nodes_io.items():
+            self.nodes_io[key].inputs = [self.uniqueNameToScopedName[node_input_id] for node_input_id in node.inputs]
+            if node.uniqueName in self.uniqueNameToScopedName:
+                self.nodes_io[key].uniqueName = self.uniqueNameToScopedName[node.uniqueName]
+
+    def to_proto(self):
+        nodes = []
+        node_stats = []
+        for v in self.nodes_io.values():
+            nodes.append(node_proto(v.uniqueName,
+                                    input=v.inputs,
+                                    outputsize=v.tensor_size,
+                                    op=v.kind,
+                                    attributes=v.attributes))
+
+            if v.tensor_size and len(v.tensor_size) > 0:  # assume data is float32, only parameter is counted
+                node_stats.append(
+                    NodeExecStats(node_name=v.uniqueName,
+                                  all_start_micros=int(time.time() * 1e7),
+                                  all_end_rel_micros=42,
+                                  memory=[AllocatorMemoryUsed(allocator_name="cpu",
+                                                              total_bytes=int(np.prod(v.tensor_size)) * 4)]))
+
+        return nodes, node_stats
+
+
+# one argument: 'hasAttribute', 'hasAttributes',
+def parse(graph, args=None, omit_useless_nodes=True):
+    """This method parses a PyTorch graph.
+    Args:
+      graph (PyTorch module): The model to be parsed.
+      args (tuple): input tensor[s] for the model.
+      omit_useless_nodes (boolean) whether to remove nodes from the graph.
+    """
+    n_inputs = len(args)  # not sure...
+
+    scope = {}
+    nodes_py = GraphPy()
+    for i, node in enumerate(graph.inputs()):
+        if omit_useless_nodes:
+            if len(node.uses()) == 0:  # number of user of the node (= number of outputs/ fanout)
+                continue
+
+        if i < n_inputs:
+            nodes_py.append(NodePyIO(node, 'input'))
+        else:
+            nodes_py.append(NodePyIO(node))  # parameter
+
+    for node in graph.nodes():
+        nodes_py.append(NodePyOP(node))
+
+    for node in graph.outputs():  # must place last.
+        NodePyIO(node, 'output')
+    nodes_py.find_common_root()
+    nodes_py.populate_namespace_from_OP_to_IO()
+    return nodes_py.to_proto()
+
+
+def graph(model, args, verbose=False, operator_export_type='ONNX', omit_useless_nodes=True):
```

Comment:
I'm confused about `omit_useless_nodes`. You run DCE optimization pass. That does the same thing; I don't see why you need another thing to do it.

Reply:
@lanpa can you answer here? Thanks.

Reply:
![image](https://user-images.githubusercontent.com/2005323/56751896-aa94cd00-67b9-11e9-8ee8-d7eb5215029e.png)
![image](https://user-images.githubusercontent.com/2005323/56752129-30187d00-67ba-11e9-9930-998ff5c5952c.png)



Some nodes `%13 : Long()` are still not eliminated.

```python
class Net1(nn.Module):
    def __init__(self):
        super(Net1, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)
        self.bn = nn.BatchNorm2d(20)

    def forward(self, x):
        x = F.max_pool2d(self.conv1(x), 2)
        x = F.relu(x) + F.relu(-x)
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = self.bn(x)
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        x = F.softmax(x, dim=1)
        return x
```


```
graph(%0 : Float(13, 1, 28, 28)
      %1 : Float(10, 1, 5, 5)
      %2 : Float(10)
      %3 : Float(20, 10, 5, 5)
      %4 : Float(20)
      %5 : Float(50, 320)
      %6 : Float(50)
      %7 : Float(10, 50)
      %8 : Float(10)
      %9 : Float(20)
      %10 : Float(20)
      %11 : Float(20)
      %12 : Float(20)
      %13 : Long()) {
  %14 : Float(13, 10, 24, 24) = onnx::Conv[dilations=[1, 1], group=1, kernel_shape=[5, 5], pads=[0, 0, 0, 0], strides=[1, 1]](%0, %1, %2), scope: Net1/Conv2d[conv1]
  %15 : Float(13, 10, 12, 12) = onnx::MaxPool[kernel_shape=[2, 2], pads=[0, 0, 0, 0], strides=[2, 2]](%14), scope: Net1
  %16 : Float(13, 10, 12, 12) = onnx::Relu(%15), scope: Net1
  %17 : Float(13, 10, 12, 12) = onnx::Neg(%15), scope: Net1
  %18 : Float(13, 10, 12, 12) = onnx::Relu(%17), scope: Net1
  %19 : Float(13, 10, 12, 12) = onnx::Add(%16, %18), scope: Net1
  %20 : Float(13, 20, 8, 8) = onnx::Conv[dilations=[1, 1], group=1, kernel_shape=[5, 5], pads=[0, 0, 0, 0], strides=[1, 1]](%19, %3, %4), scope: Net1/Dropout2d[conv2_drop]
  %21 : Float(13, 20, 4, 4) = onnx::MaxPool[kernel_shape=[2, 2], pads=[0, 0, 0, 0], strides=[2, 2]](%20), scope: Net1
  %22 : Float(13, 20, 4, 4) = onnx::Relu(%21), scope: Net1
  %23 : Float(13, 20, 4, 4) = onnx::BatchNormalization[epsilon=1e-05, momentum=1](%22, %9, %10, %11, %12), scope: Net1/BatchNorm2d[bn]
  %24 : Tensor = onnx::Constant[value=  -1  320 [ CPULongType{2} ]](), scope: Net1
  %25 : Float(13, 320) = onnx::Reshape(%23, %24), scope: Net1
  %26 : Float(320!, 50!) = onnx::Transpose[perm=[1, 0]](%5), scope: Net1/Linear[fc1]
  %27 : Float(13, 50) = onnx::Gemm[alpha=1, beta=1](%25, %26, %6), scope: Net1/Linear[fc1]
  %28 : Float(13, 50) = onnx::Relu(%27), scope: Net1
  %29 : Float(13, 50), %30 : Tensor = onnx::Dropout[ratio=0.5](%28), scope: Net1
  %31 : Float(50!, 10!) = onnx::Transpose[perm=[1, 0]](%7), scope: Net1/Linear[fc2]
  %32 : Float(13, 10) = onnx::Gemm[alpha=1, beta=1](%29, %31, %8), scope: Net1/Linear[fc2]
  %33 : Float(13, 10) = onnx::Softmax[axis=1](%32), scope: Net1
  return (%33);
}
```

Reply:
And I remove it with a heuristic condition.

```python
def parse(graph, args=None, omit_useless_nodes=True):
    import torch
    n_inputs = len(args)  # not sure...

    nodes_py = Graph_py()
    for i, node in enumerate(graph.inputs()):
        if omit_useless_nodes:
            if len(node.uses()) == 0:  # number of user of the node (= number of outputs/ fanout)
                continue
```

Reply:
I see. So the point is that DCE refuses to remove dead input arguments, because, well, that's not a valid optimization in general. But where did node 13 come from in the first place? Sounds like some sort of tracing bug... (This comment does not require action for experimental merge.)